### ISSUE TITLE
Allied supply network merging doesn't add connetions between two empires' networks for a third mutually allied empire

### ISSUE THREAD

**Issue by magnate** [freeorion/freeorion#2905]
### Environment
* **FreeOrion Version**: b2112fe
* **Operating System**: Win7
* **Graphic card used**: Radeon 5750
* **Fetched as** 
  * Binary release
  * Weekly development build

### Description
In the just-finished 9th MP game, one ally gifted me a planet with an Enclave Of The Void and another gifted me one with a Black Hole Power Generator. The bonus from the former was immediately visible, from the turn I got it. All my research planets benefitted from it. 

The bonus from the later never appeared. I think the reason why is because the Enclave was directly connected to my supply, while the BHG was further west, with an ally in between. So while it was connected to me via allied supply, it wasn't directly connected. I'm not sure if this is a bug or WAD. 

### Expected Result
I would expect all allied supply to be equal, so the BHG bonus should have been given to my industry planets. 

### Steps to reproduce
1. Create a building that provides a bonus, such as an Enclave or a BHG.
2. Gift it to your ally
3. If your ally is directly supply connected to you, s/he will see the bonus. If the recipient is two allies away, s/he won't. I do not know if this also applies to weapon upgrades reaching your ships in allied space.

**Comment by Oberlus** [comment]
There is no allied supply as such. Alliances does not form a single supply group and hence can't share things like species or building bonuses.
If you see in the map that one of your planets is in a different different supply group, "connected" by the supply group of an ally, you have two disconnected supply groups.
Changing this by making a single supply group for all the allies in an alliance is hard because it would imply many nasty things:
- A single building of a kind would be enough to grant the corresponding boost to all the allied empires.
- PP are shared among all planets in the same supply group, and there are no rules to manage how to distribute the PP produced by each empire among the items in production on each empire's production queue.

Keeping separate supply groups for each empire in the alliance while propagating PP and whatnot through the supply of allied empires to connect different supply groups of the same empire is not supported. That would be the fix for this issue. But it is not simple to implement.

**Comment by magnate** [comment]
Ah, so I have misunderstood. The Enclave bonus happened because it doesn't require a supply connection, right? And the BHG bonus wouldn't happen in any allied territory, because it would not be supply connected. Sorry, this should be rebadged as a feature request for allied supply then.

**Comment by magnate** [comment]
Btw I don't think you need to allow a single building to grant its bonuses to all connected allies. You can still require the building to be owned by the empire getting the bonus.

**Comment by Oberlus** [comment]
> Btw I don't think you need to allow a single building to grant its bonuses to all connected allies. You can still require the building to be owned by the empire getting the bonus.

I meant that if you change backend to consider all empires in alliance to share a single supply group, without changing anything else, then bonus from a single building would apply to all empires in that supply group. I agree that sounds bad.
So the way to go is to allow the fragmented supply groups of a single empire to be considered as connected by the backend when they are "connected" by an ally's supply. This needs delicate work in the backend and does not seem to gonna happen any time soon.

**Comment by magnate** [comment]
Ok that's fair enough, I can imagine it's more difficult than it sounds. Yes I agree that having a single alliance supply would be bad because of the one building thing.

**Comment by geoffthemedio** [comment]
> So the way to go is to allow the fragmented supply groups of a single empire to be considered as connected by the backend when they are "connected" by an ally's supply. This needs delicate work in the backend and does not seem to gonna happen any time soon.

For purposes of pooling and distributing a single empire's PP, this should already happen: two separate supply groups of the same empire, connected to eachother by the supply network of an allied empire, using PP output from one of the two supply groups to produce something in the other supply group. There have been bugs reported and resolved about corner cases where it didn't work, eg. #2554

I'm not entirely sure, but based on #2500 I think an allied supply network should also connect two separate supply groups (of the same empire, via the lanes of the other empire's adjacent supply network) for purposes of whether a location in one of the supply groups is supply-connected to a location in the other supply group, ie. in a `ResourceSupplyConnected` condition.

These are separate issues to whether a building or special on a planet owned by one empire can be supply connected to planets of another empire, or whether production output from one empire can be spent by another empire to produce something, both of which should presently not be possible, as far as I recall.

**Comment by magnate** [comment]
> I'm not entirely sure, but based on #2500 I think an allied supply network should also connect two separate supply groups (of the same empire, via the lanes of the other empire's adjacent supply network) for purposes of whether a location in one of the supply groups is supply-connected to a location in the other supply group, ie. in a `ResourceSupplyConnected` condition.

Ok, this is the bug I was originally reporting. If this is supposed to be working, it's not working when there are two allied supply networks connecting my two supply groups. It may be working when there is only one, the case I noted didn't show that either way.

**Comment by geoffthemedio** [comment]
> there are two allied supply networks connecting my two supply groups

That means: a path from your supply network to your other supply network must pass through two different allies' supply networks?

**Comment by magnate** [comment]
That's right. My assumption was that it was implemented/fixed for one ally's supply network, but not for two.

**Comment by geoffthemedio** [comment]
That was not at all clear from your original post:
> the BHG was further west, with **_an ally in between_**.

But it's plausible that it won't work for allied supply connections that must pass through a total of three allied empires' supply networks. The comments in the C++ code suggest that it finds cases where your empire has an obstructed supply network connection between two systems where an allied empire also has an obstructed supply network connection going in the opposite direction. In that case, it adds supply connections for both empires between those systems. It also adds all of an allied empire's internal supply network connections to your empire. But it doesn't add supply network connections to your empire in cases where two of your allies have obstructed supply network connections in opposite directions between two systems where you don't have any supply connection. So in that case, you can propagate supply into an ally's supply network, but not from one ally's network into another ally's network, regardless of whether or not those allies (of yours) are also allied.

**Comment by magnate** [comment]
My apologies for the lack of clarity, it's hard to describe! Of course, I was describing a team game where all are allied. You are right that it's more complex if my two allies are not allied with each other! Thank you for clarifying.
